File: LeetCode/1745. Palindrome Partitioning IV/solution.py
class Solution:
    def checkPartitioning(self, s: str) -> bool:
        n = len(s)
        dp = [[False]*n for _ in range(n)]
        for i in range(n):
            dp[i][i] = True
            if i + 1 < n:
                dp[i+1][i] = True
        for k in range(1,n):
            for i in range(n-k):
                dp[i][i+k] = dp[i+1][i+k-1]&(s[i]==s[i+k])

        for i in range(1,n-1):
            for j in range(i+1,n):
                if dp[0][i-1] and dp[i][j-1] and dp[j][n-1]:
                    return True
        return False 

File: LeetCode/1745. Palindrome Partitioning IV/solution.cpp
const int maxn = 2005;
bitset<maxn> dp[maxn];

class Solution {
public:
    bool checkPartitioning(string s) {
        int n = s.length();
        for (int i = 0; i < n; i++)
            dp[i].set();
        for (int i = n-1; i >= 0; i--){
            for (int j = i+1; j < n; j++)
                if (dp[i+1][j-1] && s[i] == s[j])
                    dp[i][j] = 1;
                else
                    dp[i][j] = 0;
        }
        // for (int i = 0; i < n; i++)
        //     cout << dp[i] << endl;
        for (int i = 1; i < n; i++){
            for (int j = i; j < n-1; j++)
                if (dp[0][i-1] && dp[i][j] && dp[j+1][n-1])
                    return true;
        }
        // for (int i = 0; i < n; i++)
        //     cout << dp[i] << endl;
        return false;

    }
};